### PR TITLE
chore: remove old hyper and bump prost and tonic versions

### PR DESCRIPTION
**Pull request by stedfn**

### CODE REVIEW

**Comment by venkkatesh-sekar** [comment]
> Vulnerable dependency information

This was due to a mixup in testing. Should be resolved now.

**Review by NikolasHai — APPROVED** [review]

File: Cargo.Bazel.toml.lock [hunk 1]
```diff
@@ -3220,8 +3219,9 @@ dependencies = [
  "prometheus-parse",
  "proptest",
  "proptest-derive",
- "prost",
- "prost-build",
+ "prost 0.12.2",
+ "prost 0.13.2",
+ "prost-build 0.13.2",
```
Comment (nmattia) [inline Cargo.Bazel.toml.lock:3224]: This seems to duplicate dependencies. Can you check why both `0.12.2` & `0.13.2` are in use?

**Review by randombit — APPROVED** [review]

File: rs/nervous_system/proto/protobuf_generator/src/lib.rs [hunk 2]
```diff
@@ -46,13 +38,7 @@ pub fn generate_prost_files(proto_paths: ProtoPaths<'_>, out_dir: &Path) {
         .nervous_system
         .join("ic_nervous_system/pb/v1/nervous_system.proto");
 
-    // Most types should be Copy (currently, only Image is not Copy).
-    for type_name in copy_type_names {
```
Comment (max-dfinity) [inline rs/nervous_system/proto/protobuf_generator/src/lib.rs:50]: Why this change to the generator?  This would mean that all of our proto types are now copy, which is not necessarily desirable

File: rs/nervous_system/proto/protobuf_generator/src/lib.rs [hunk 3]
```diff
@@ -14,14 +14,6 @@ impl ProtoPaths<'_> {
 
 /// Build protos using prost_build.
 pub fn generate_prost_files(proto_paths: ProtoPaths<'_>, out_dir: &Path) {
-    let copy_type_names = vec![
```
Comment (max-dfinity) [inline rs/nervous_system/proto/protobuf_generator/src/lib.rs:17]: Why are we removing copy from these?

File: Cargo.Bazel.toml.lock (hunk 1 above)
Comment (stedfn) [reply Cargo.Bazel.toml.lock:3224]: [pprof](https://docs.rs/pprof/latest/pprof/) depends on prost 0.12. It has not been upgraded yet

**Review by fspreiss — APPROVED** [review]

File: rs/replicated_state/src/metadata_state.rs [hunk 4]
```diff
@@ -1299,7 +1299,7 @@ impl TryFrom<pb_queues::Stream> for Stream {
             .map(|signal| {
                 Ok(RejectSignal {
                     reason: pb_queues::RejectReason::try_from(signal.reason)
-                        .map_err(ProxyDecodeError::DecodeError)?
```
Comment (rumenov) [inline rs/replicated_state/src/metadata_state.rs:1302]: @stedfn why did we change this ?

File: rs/sns/governance/protobuf_generator/src/lib.rs [hunk 5]
```diff
@@ -44,7 +44,7 @@ pub fn generate_prost_files(proto: ProtoPaths<'_>, out: &Path) {
         "ic_sns_governance.pb.v1.NeuronId",
         "#[derive(Eq, std::hash::Hash)]",
     );
-    config.type_attribute("ic_sns_governance.pb.v1.ProposalId", "#[derive(Eq, Copy)]");
+    // config.type_attribute("ic_sns_governance.pb.v1.ProposalId", "#[derive(Eq, Copy)]");
```
Comment (rumenov) [inline rs/sns/governance/protobuf_generator/src/lib.rs:47]: maybe remove

**Review by rumenov — APPROVED** [review]

File: rs/replicated_state/src/metadata_state.rs (hunk 4 above)
Comment (rumenov) [reply rs/replicated_state/src/metadata_state.rs:1302]: ok, the reason is because the return type of try_from is different from before, namely it returns [prost::UnknownEnumValue](https://docs.rs/prost/0.13.3/prost/struct.UnknownEnumValue.html)

**Review by dsarlis — APPROVED** [review]

**Review by tthebst — APPROVED** [review]

File: Cargo.Bazel.toml.lock (hunk 1 above)
Comment (nmattia) [reply Cargo.Bazel.toml.lock:3224]: can we upgrade it then? or does this cause failures?

Comment (stedfn) [reply Cargo.Bazel.toml.lock:3224]: The latest version of pprof uses 0.12, so it can't be upgraded now. It should be easy once the new version is released

Comment (nmattia) [reply Cargo.Bazel.toml.lock:3224]: Gotcha. Can we wait for the new pprof version to be released then or is there some urgency in bumping those deps now?

**Review by alin-at-dfinity — APPROVED** [review]

File: Cargo.Bazel.toml.lock (hunk 1 above)
Comment (stedfn) [reply Cargo.Bazel.toml.lock:3224]: Ideally not as this is my final day at DFINITY and it is also a step in the right direction regarding cleaning up dependencies which will eventually allow us to not pull the old hyper at all

Comment (rumenov) [reply Cargo.Bazel.toml.lock:3224]: We can't wait for pprof. They upgrade their crate fairly slow. 

Again we should not worry to much about transitive dependencies of our external crates as long as use external crates from credible organisations.

Comment (rumenov) [reply Cargo.Bazel.toml.lock:3225]: For the foreseeable future there always can be multiple versions due to transitive dependencies. This is less of a problem than keeping our direct dependencies stale.

**Review by nmattia — APPROVED** [review]

File: rs/sns/governance/src/proposal.rs [hunk 6]
```diff
@@ -2168,7 +2168,7 @@ impl ProposalData {
         let limited_ballots: BTreeMap<_, _> = ballots
             .iter()
             .filter(|(neuron_id, _)| caller_neurons_set.contains(*neuron_id))
-            .map(|(neuron_id, ballot)| (neuron_id.clone(), ballot.clone()))
+            .map(|(neuron_id, ballot)| (neuron_id.clone(), *ballot))
```
Comment (aterga) [inline rs/sns/governance/src/proposal.rs:2171]: Is NeuronId not copy?

**Review by aterga — APPROVED** [review]

File: rs/sns/governance/src/proposal.rs (hunk 6 above)
Comment (max-dfinity) [reply rs/sns/governance/src/proposal.rs:2171]: It was not copy, now it is copy, b/c of prost change (I learned this in response to a similar question)

**Review by max-dfinity — APPROVED** [review]

File: rs/sns/governance/src/types.rs [hunk 7]
```diff
@@ -977,6 +976,15 @@ impl From<prost::DecodeError> for GovernanceError {
     }
 }
 
+impl From<prost::UnknownEnumValue> for GovernanceError {
```
[the remark this answers is not in the document]
Comment (rumenov) [reply rs/sns/governance/src/types.rs:979]: CanisterInstallMode::try_from returns prost::UnknownEnumValue now and not prost::DecodeError like before
